### DISCUSSION TITLE
Consider Supporting Exporting Playlist to M3U

### DISCUSSION THREAD

**Discussion by rstrube** [Taiko2k/Tauon#639] [Ideas]
I love TauonMusicBox and am a sponsor of the project.

I use the sync playlist to copy my music to my Android device which works wonderfully.  The app I use for offline music playback is called Musicolet and is a great Android music player and manager.

Unfortunately Musicolet only supports playlists in the M3U format, and Tauon only exports playlists to XSPF.  I've been able to use VLC to open the XSPF file, save it as an M3U and then perform some manual cleanup, but this is quite time consuming.

It would be really nice/convenient to support exporting playlists directly to M3U and I would really appreciate if this feature was considered.

Thanks!
Rob

**Comment by Taiko2k** [comment]
Could you help me understand how filepaths work in M3U? Or what works / what would be expected? Like, if it were to work like the existing *export XSPF* function, the paths would be full paths to where they are on your PC, does that work on your Android device as is? Also what cleanup are you preforming?

**Comment by rstrube** [comment]
Hi @Taiko2k !

Very sorry for the late response.  You are correct that M3U could _potentially_ contain file-paths that are unique to my PC (for example `/home/robert/Music/xxx/xxx`).  But this largely depends on where the M3U file is relative to the music.  For example an M3U file located in my home folder could look like this and still be valid.  All paths would be relative to the location of the M3U file itself.

```
#EXTM3U
#EXTINF:154,The Four Horsemen - No Lie
Music/The%20Four%20Horsemen/Something%20True/03%20No%20Lie.mp3
#EXTINF:222,ELFL - Purple Voyager
Music/ELFL/Purple%20Voyager/01%20Purple%20Voyager.mp3
#EXTINF:238,Foals - My Number
Music/Foals/Holy%20Fire/03%20My%20Number.mp3
```

You can also have absolute file paths, something like this:

```
#EXTM3U
#EXTINF:154,The Four Horsemen - No Lie
file:///home/robert/Music/The%20Four%20Horsemen/Something%20True/03%20No%20Lie.mp3
#EXTINF:222,ELFL - Purple Voyager
file:///home/robert/Music/ELFL/Purple%20Voyager/01%20Purple%20Voyager.mp3
#EXTINF:238,Foals - My Number
file:///home/robert/Music/Foals/Holy%20Fire/03%20My%20Number.mp3
```

In that case I would need to use sed or awk to replace the: `/home/robert` with `/storage/emulated/0` which matches up to the home folder on my Android device.

The main challenge that I'm facing is that VLC will allow you to open up an XSPF playlist file and save it as an M3U file, but there is a bug in their implementation.  The metadata information it generates is corrupted and somehow causes Musicolet on Android to fail and import the M3U file.

For example VLC will generate a M3U like this:

```
#EXTM3U
#EXTINF:154,The Four Horsemen - No Lie
Music/The%20Four%20Horsemen/Something%20True/03%20No%20Lie.mp3
#EXTINF:222,ELFL - 01 Purple Voyager.mp3
Music/ELFL/Purple%20Voyager/01%20Purple%20Voyager.mp3
#EXTINF:238,Foals - 03 My Number.mp3
Music/Foals/Holy%20Fire/03%20My%20Number.mp3
```

Notice that the metadata lines (for any track past the first track) actually include the file name.
* `#EXTINF:222,ELFL - 01 Purple Voyager.mp3`
* `#EXTINF:238,Foals - 03 My Number.mp3`

These should actually be:
* `#EXTINF:222,ELFL - Purple Voyager`
* `#EXTINF:238,Foals - My Number`

My solution for now is to simply remove all the metadata lines:

```
#EXTM3U
Music/The%20Four%20Horsemen/Something%20True/03%20No%20Lie.mp3
Music/ELFL/Purple%20Voyager/01%20Purple%20Voyager.mp3
Music/Foals/Holy%20Fire/03%20My%20Number.mp3
```
I can do this in vim fairly easily, but it's still a pain as I have to:

1. Export the playlist to XSPF
2. Open playlist in VLC
3. Save it as an M3U file
4. Open in vim and manually remove all the corrupted metadata lines

Luckily Musicolet will follow the filepaths from the M3U file and just pull the metadata from the MP3 file itself.

It would be fantastic if you could implement a feature to just dump out an M3U file.  I'm not sure if you would want to do absolute pathing or relative pathing (absolute would probably be preferable from the application's perspective).  Even without the extra metadata lines it would be very beneficial.

Let me know if you have any other questions.  I'm a developer myself but I focus on web applications - Javascript, C#, React, Node.js, etc. I might be able to assist if you think there's an overlap in the skills required!

**Reply by Taiko2k** [reply]
Just dumping an m3u with absolute paths would be far simpler and pretty easy to implement. I'll plan to get that done for the next version of Tauon.

**Comment by rstrube** [comment]
Thank you very much for consideration of this feature.  I've found some additional information on the M3U format here: https://en.wikipedia.org/wiki/M3U

I was confused about the numeric value that appears in the `#EXTINF` lines.  Turns out this is the length of the audio in seconds.

So:

```
#EXTINF:154,The Four Horsemen - No Lie
Music/The%20Four%20Horsemen/Something%20True/03%20No%20Lie.mp3
```

Would equate to a song that is 154 seconds in length.

**Comment by da2x** [comment]
The minimum effort M3U8 file (UTF-8 version of M3U) is just:

```
#EXTM3U
/home/Music/Artist/Song 1.oga
/home/Music/Artist/Song 2.oga
```

It’s accepted by iTunes and VLC.

**Reply by Taiko2k** [reply]
Both import and export of M3U should be more or less working already. 

To export, right click playlist tab > export... > enter a path, tick the M3U box and click export.
